blockingconcurrentqueue.h

```diff
@@ -519,7 +551,7 @@ class BlockingConcurrentQueue
 	inline bool enqueue(T const& item)
```

Comment:
Shouldn't this reduce the count of the semaphore?

Reply:
Well, my thinking here was that enqueue() must never fail, so in effect this would probably increase the amount of enqueue-slots available in the underlying buffer.
By not decreasing the count here, it will implicitly increase the enqueue semaphore later when the slot becomes free again from dequeueing, accounting for the extra slot that became available in the queue.

Although this line of thinking is a bit weird.
I think I'll probably instead just decrease the semaphore; I'll have to unconditionally decrease beyond zero though to not destroy the never-fails property of enqueue()

Reply:
Ah, I think I understand, at least the first part.
Unfortunately, `enqueue` can still fail for various reasons, e.g. if there's not enough memory. But even if it succeeds, it does not necessarily create a new slot -- it tries to use an existing block, if any, then requests a new block from the queue if that fails, which may give it a recycled or preallocated block. A new block is allocated only if that also fails. (And a block has more than one slot.)

So if there's one call the `enqueue` and another to `try_enqueue`, the `try_enqueue` might fail if the `enqueue` took the last slot even though if called in the opposite order they would both succeed since `try_enqueue` would end up allocating a new block.